File: src/shipyard/executor/clixml.py
"""Decode PowerShell CLIXML error envelopes surfaced by remote
Windows commands over SSH.

When a Windows PowerShell session writes to any non-stdout stream
(``Write-Error``, uncaught exceptions, ``Write-Warning``, etc.) and
the stream is relayed back to the SSH client, the receiving side
sees a payload that starts with the literal bytes ``#< CLIXML``
followed by a PowerShell CLIXML (Common Language Infrastructure
XML) document. The actual human-readable diagnostic — "file in
use," "git exit 1," "missing dependency," etc. — is buried inside
``<S S="Error">…</S>`` elements in that XML, not in plain text.

Pre-decoder, shipyard's summary row looked like::

    ✗ windows: Bundle apply failed: Remote bundle apply failed: #< CLIXML

Which is exactly as useful as the bare ``windows  error  ssh-windows``
row we fixed in v0.28.0. This module closes that gap (#188): if the
stderr buffer starts with the CLIXML sentinel, we extract the error
and warning stream text, decode PowerShell's ``_x000D_``/``_x000A_``
escape sequences back to CR/LF, and hand the caller plain text that
names the actual failure.

Malformed or partially-transferred envelopes fall through to the
original raw text — the pre-decoder behaviour — so we never make a
failure case worse.
"""

from __future__ import annotations

import re
from xml.etree import ElementTree as ET

# PowerShell's CLIXML stream-header sentinel. Case-insensitive; older
# hosts sometimes send the header with a different whitespace between
# `#<` and `CLIXML`. Any stderr buffer whose first non-whitespace
# characters match this gets routed through `_decode`.
_CLIXML_SENTINEL = "#< CLIXML"

# Powershell serializes the XML 1.0 "restricted" character set by
# encoding forbidden code points as ``_xHHHH_`` (four-digit hex).
# Most commonly ``_x000D_`` / ``_x000A_`` / ``_x0009_`` for CR / LF /
# tab in multi-line error messages. Decode them back so the printed
# text preserves readable line breaks.
_ESCAPE_RE = re.compile(r"_x([0-9A-Fa-f]{4})_")

# Cap the decoded payload so a multi-kilobyte traceback doesn't
# balloon the terminal when surfaced beneath the summary table. The
# per-target log file still carries the full envelope.
_MAX_DECODED_CHARS = 800


def is_clixml(text: str) -> bool:
    """True iff ``text`` starts with a CLIXML envelope header."""
    return text.lstrip().startswith(_CLIXML_SENTINEL)


def maybe_decode_clixml(text: str) -> str:
    """Best-effort: if ``text`` looks like a CLIXML envelope, return
    the decoded human-readable error. Otherwise return ``text`` as-is.

    Never raises — malformed CLIXML falls back to the original text
    so callers' existing error surfaces keep working on partial or
    non-standard envelopes.
    """
    if not is_clixml(text):
        return text
    try:
        decoded = _decode(text)
    except (ET.ParseError, ValueError):
        return text
    return decoded or text


def _decode(text: str) -> str:
    # Strip everything up to and including the sentinel; the rest
    # should be a well-formed XML document.
    idx = text.find(_CLIXML_SENTINEL)
    xml_blob = text[idx + len(_CLIXML_SENTINEL):].lstrip()

    # The stream can contain multiple concatenated ``<Objs>…</Objs>``
    # documents back-to-back (one per stderr flush). Parse each
    # independently so a malformed trailing doc doesn't throw away
    # an earlier well-formed one.
    messages: list[str] = []
    for obj_doc in _split_objs(xml_blob):
        messages.extend(_extract_messages(obj_doc))
    if not messages:
        return ""

    # Preserve order, dedupe consecutive exact-same lines (PowerShell
    # tends to emit both a short ``Message`` and a longer
    # ``FullyQualifiedErrorId`` derived from the same exception).
    deduped: list[str] = []
    for m in messages:
        stripped = m.strip()
        if not stripped:
            continue
        if deduped and deduped[-1] == stripped:
            continue
        deduped.append(stripped)

    joined = "\n".join(deduped)
    if len(joined) > _MAX_DECODED_CHARS:
        # Keep the tail — last errors in a multi-line cascade are
        # almost always the proximate cause.
        joined = "…" + joined[-(_MAX_DECODED_CHARS - 1):]
    return joined


def _split_objs(xml_blob: str) -> list[str]:
    """Split a concatenation of ``<Objs>…</Objs>`` documents into a
    list of individual XML strings. Returns ``[xml_blob]`` unchanged
    when only one document is present; empty list when the blob is
    empty or shaped wrong for recovery."""
    out: list[str] = []
    start = 0
    while start < len(xml_blob):
        open_idx = xml_blob.find("<Objs", start)
        if open_idx < 0:
            break
        close_idx = xml_blob.find("</Objs>", open_idx)
        if close_idx < 0:
            break
        end = close_idx + len("</Objs>")
        out.append(xml_blob[open_idx:end])
        start = end
    return out


def _extract_messages(obj_doc: str) -> list[str]:
    """Pull the human-readable text out of a single ``<Objs>``
    document. Looks at both ``<S S="...">`` stream records (the
    common case: ``Write-Error`` / ``Write-Warning`` output) and
    nested ``<Obj>`` records with ``<S N="Message">`` /
    ``<S N="Exception">`` properties (full ``ErrorRecord``
    serializations)."""
    root = ET.fromstring(obj_doc)

    # PowerShell CLIXML uses a default namespace on ``<Objs>``. ElementTree
    # preserves that as ``{ns}tag`` when walking, so match on local
    # names without hardcoding the namespace URI.
    results: list[str] = []
    for elem in root.iter():
        tag = _local_name(elem.tag)
        if tag == "S":
            # Stream record: S="Error" / S="Warning" / S="Verbose" /
            # etc. Or property record: N="Message" / N="Exception".
            stream = elem.get("S", "")
            prop_name = elem.get("N", "")
            if (
                stream in {"Error", "Warning"}
                or prop_name in {"Message", "Exception", "FullyQualifiedErrorId"}
            ):
                results.append(_decode_escapes(elem.text or ""))
    return results


def _local_name(tag: str) -> str:
    # Strip ``{ns}`` prefix from ``ElementTree.Element.tag``.
    if "}" in tag:
        return tag.split("}", 1)[1]
    return tag


def _decode_escapes(raw: str) -> str:
    """Decode PowerShell's ``_xHHHH_`` escapes back to real chars."""
    def _sub(m: re.Match[str]) -> str:
        try:
            return chr(int(m.group(1), 16))
        except ValueError:
            return m.group(0)
    return _ESCAPE_RE.sub(_sub, raw)

File: src/executor/clixml.rs
//! Best-effort PowerShell CLIXML stderr decoding.
//!
//! Windows PowerShell can serialize non-stdout streams as CLIXML when
//! relayed over SSH. This decoder extracts human-readable error and
//! warning text while preserving the original text on malformed input.

const CLIXML_SENTINEL: &str = "#< CLIXML";
const MAX_DECODED_CHARS: usize = 800;

/// Return true when `text` starts with a CLIXML stream header after
/// leading whitespace.
#[must_use]
pub fn is_clixml(text: &str) -> bool {
    text.trim_start().starts_with(CLIXML_SENTINEL)
}

/// Decode a CLIXML stderr payload when possible.
///
/// Text that appears before the first CLIXML sentinel is preserved and
/// joined with decoded envelope messages. If no sentinel exists, or if
/// the envelope is malformed enough that no useful text can be recovered,
/// the original text is returned unchanged.
#[must_use]
pub fn maybe_decode_clixml(text: &str) -> String {
    if !is_clixml(text) && !text.contains(CLIXML_SENTINEL) {
        return text.to_owned();
    }

    let (prefix, xml_blob) = split_once_sentinel(text);
    let messages = extract_messages_from_blob(xml_blob);
    let decoded = dedupe_and_limit(messages);
    let parts = [prefix.trim(), decoded.trim()]
        .into_iter()
        .filter(|part| !part.is_empty())
        .collect::<Vec<_>>();

    if parts.is_empty() {
        text.to_owned()
    } else {
        parts.join("\n")
    }
}

fn split_once_sentinel(text: &str) -> (&str, &str) {
    let Some(index) = text.find(CLIXML_SENTINEL) else {
        return (text, "");
    };
    let prefix = &text[..index];
    let blob = text[index + CLIXML_SENTINEL.len()..].trim_start();
    (prefix, blob)
}

fn extract_messages_from_blob(xml_blob: &str) -> Vec<String> {
    split_objs(xml_blob)
        .into_iter()
        .flat_map(extract_messages)
        .collect()
}

fn split_objs(xml_blob: &str) -> Vec<&str> {
    let mut documents = Vec::new();
    let mut start = 0;

    while start < xml_blob.len() {
        let Some(open_rel) = xml_blob[start..].find("<Objs") else {
            break;
        };
        let open = start + open_rel;
        let Some(close_rel) = xml_blob[open..].find("</Objs>") else {
            break;
        };
        let close = open + close_rel + "</Objs>".len();
        documents.push(&xml_blob[open..close]);
        start = close;
    }

    documents
}

fn extract_messages(document: &str) -> Vec<String> {
    let mut messages = Vec::new();
    let mut start = 0;

    while let Some(tag_rel) = document[start..].find("<S") {
        let tag_start = start + tag_rel;
        let Some(tag_end_rel) = document[tag_start..].find('>') else {
            break;
        };
        let tag_end = tag_start + tag_end_rel;
        let tag = &document[tag_start..=tag_end];
        let content_start = tag_end + 1;
        let Some(close_rel) = document[content_start..].find("</S>") else {
            break;
        };
        let content_end = content_start + close_rel;

        if is_message_tag(tag) {
            messages.push(decode_powershell_escapes(&decode_xml_entities(
                &document[content_start..content_end],
            )));
        }

        start = content_end + "</S>".len();
    }

    messages
}

fn is_message_tag(tag: &str) -> bool {
    tag.contains(r#"S="Error""#)
        || tag.contains(r#"S="Warning""#)
        || tag.contains(r#"N="Message""#)
        || tag.contains(r#"N="Exception""#)
        || tag.contains(r#"N="FullyQualifiedErrorId""#)
}

fn decode_xml_entities(raw: &str) -> String {
    raw.replace("&lt;", "<")
        .replace("&gt;", ">")
        .replace("&quot;", "\"")
        .replace("&apos;", "'")
        .replace("&amp;", "&")
}

fn decode_powershell_escapes(raw: &str) -> String {
    let mut output = String::with_capacity(raw.len());
    let mut index = 0;

    while index < raw.len() {
        let rest = &raw[index..];
        if rest.starts_with("_x") && rest.len() >= 7 && &rest[6..7] == "_" {
            let hex = &rest[2..6];
            if let Ok(value) = u32::from_str_radix(hex, 16)
                && let Some(ch) = char::from_u32(value)
            {
                output.push(ch);
                index += 7;
                continue;
            }
        }
        let Some(ch) = rest.chars().next() else {
            break;
        };
        output.push(ch);
        index += ch.len_utf8();
    }

    output
}

fn dedupe_and_limit(messages: Vec<String>) -> String {
    let mut deduped: Vec<String> = Vec::new();

    for message in messages {
        let stripped = message.trim();
        if stripped.is_empty() {
            continue;
        }
        if deduped.last().is_some_and(|previous| previous == stripped) {
            continue;
        }
        deduped.push(stripped.to_owned());
    }

    let joined = deduped.join("\n");
    if joined.chars().count() <= MAX_DECODED_CHARS {
        return joined;
    }

    let tail = joined
        .chars()
        .rev()
        .take(MAX_DECODED_CHARS - 1)
        .collect::<Vec<_>>()
        .into_iter()
        .rev()
        .collect::<String>();
    format!("...{tail}")
}

#[cfg(test)]
mod tests {
    use super::{is_clixml, maybe_decode_clixml};

    #[test]
    fn plain_text_without_sentinel_is_unchanged() {
        assert_eq!(
            maybe_decode_clixml("just a regular error"),
            "just a regular error"
        );
    }

    #[test]
    fn detects_clixml_after_leading_whitespace() {
        assert!(is_clixml("  #< CLIXML\n<Objs></Objs>"));
        assert!(!is_clixml("error\n#< CLIXML\n<Objs></Objs>"));
    }

    #[test]
    fn decodes_error_stream_text() {
        let envelope = concat!(
            "#< CLIXML\n",
            r#"<Objs xmlns="http://schemas.microsoft.com/powershell/2004/04">"#,
            r#"<S S="Error">inner PS error</S>"#,
            "</Objs>"
        );
        assert_eq!(maybe_decode_clixml(envelope), "inner PS error");
    }

    #[test]
    fn decodes_message_properties_and_xml_entities() {
        let envelope = concat!(
            "#< CLIXML\n",
            "<Objs>",
            r#"<S N="Message">bad &lt;path&gt; &amp; reason</S>"#,
            "</Objs>"
        );
        assert_eq!(maybe_decode_clixml(envelope), "bad <path> & reason");
    }

    #[test]
    fn decodes_powershell_restricted_character_escapes() {
        let envelope = concat!(
            "#< CLIXML\n",
            "<Objs>",
            r#"<S S="Error">line1_x000D__x000A_line2_x0009_tab</S>"#,
            "</Objs>"
        );
        assert_eq!(maybe_decode_clixml(envelope), "line1\r\nline2\ttab");
    }

    #[test]
    fn surfaces_pre_sentinel_stderr_with_progress_only_body() {
        let envelope = concat!(
            "error: could not open 'C:/Users/alice/shipyard.bundle'\n",
            "#< CLIXML\n",
            r#"<Objs Version="1.1.0.1">"#,
            r#"<Obj S="progress"><TN><T>PSCustomObject</T></TN></Obj>"#,
            "</Objs>"
        );
        let decoded = maybe_decode_clixml(envelope);
        assert!(decoded.contains("could not open"));
        assert!(!decoded.contains("#< CLIXML"));
        assert!(!decoded.contains("<Objs"));
    }

    #[test]
    fn joins_pre_sentinel_stderr_and_error_stream() {
        let envelope = concat!(
            "error: outer context\n",
            "#< CLIXML\n",
            r#"<Objs xmlns="http://schemas.microsoft.com/powershell/2004/04">"#,
            r#"<S S="Error">inner PS error</S>"#,
            "</Objs>"
        );
        assert_eq!(
            maybe_decode_clixml(envelope),
            "error: outer context\ninner PS error"
        );
    }

    #[test]
    fn consecutive_duplicate_messages_are_deduped() {
        let envelope = concat!(
            "#< CLIXML\n",
            "<Objs>",
            r#"<S S="Error">same</S>"#,
            r#"<S N="Message">same</S>"#,
            r#"<S S="Warning">different</S>"#,
            "</Objs>"
        );
        assert_eq!(maybe_decode_clixml(envelope), "same\ndifferent");
    }
}
